isles/scoring.py

```diff
@@ -331,4 +331,156 @@ def accuracy(truth, prediction, batchwise=False):
                                        prediction[sample_idx,...].reshape((num_pred,)),
                                        batchwise=False)
             accuracy_list.append(sample_accuracy)
-        return tuple(accuracy_list)
\ No newline at end of file
+        return tuple(accuracy_list)
+
+
+def compute_lesion_f1_score(truth, prediction, empty_value=1.0, connectivity=26):
+    """
+    Computes the lesion-wise F1-score between two masks.
+
+    Parameters
+    ----------
+    ground_truth : array-like, bool
+        Any array of arbitrary size. If not boolean, will be converted.
+    prediction : array-like, bool
+        Any other array of identical size as 'ground_truth'. If not boolean, it will be converted.
+    empty_value : scalar, float.
+    connectivity : scalar, int.
+
+    Returns
+    -------
+    f1_score : float
+        Lesion-wise F1-score as float.
+        Max score = 1
+        Min score = 0
+        If both images are empty (tp + fp + fn =0) = empty_value
+
+    Notes
+    -----
+    This function computes lesion-wise score by defining true positive lesions (tp), false positive lesions (fp) and
+    false negative lesions (fn) using 3D connected-component-analysis.
+
+    tp: 3D connected-component from the ground-truth image that overlaps at least on one voxel with the prediction image.
+    fp: 3D connected-component from the prediction image that has no voxel overlapping with the ground-truth image.
+    fn: 3d connected-component from the ground-truth image that has no voxel overlapping with the prediction image.
+    """
+    truth = np.asarray(truth).astype(bool)
+    prediction = np.asarray(prediction).astype(bool)
+    tp = 0
+    fp = 0
+    fn = 0
+
+    # Check if ground-truth connected-components are detected or missed (tp and fn respectively).
+    intersection = np.logical_and(truth, prediction)
+    labeled_ground_truth, N = cc3d.connected_components(truth, connectivity=connectivity, return_N=True)
+
+    # Iterate over ground_truth clusters to find tp and fn.
+    # tp and fn are only computed if the ground-truth is not empty.
+    if N > 0:
```

Comment:
I don't think `N` is needed. `cc3d.each` returns an empty iterator if `N==0`, so it would skip the section anyway.